Runtime error: invalid memory address or nil pointer dereference

LND stopped. I'm not sure if there's much to do about this or anything but it's certainly unexpected/undesired. I'm hoping the shutdown logs may help to fix an issue. Feel free to close this one if it's not helpful.

Version 0.4, on a Debian 4.9.20-grsec, 25G free RAM, Bitcoin Core RPC client version v0.15.99.0-9cf6393a4

2018-03-29 04:49:24.813 [INF] CRTR: Block 000000000000000000155720ac9c425fc94b9c6343e2da0a4bb43dbb735544e0 (height=515624) clo
sed 0 channels
2018-03-29 04:49:24.867 [INF] NTFN: New block: height=515624, sha=000000000000000000155720ac9c425fc94b9c6343e2da0a4bb43dbb7355
44e0
2018-03-29 04:49:24.867 [INF] UTXN: Attempting to graduate height=515624: num_kids=0, num_babies=0
2018-03-29 04:49:26.692 [INF] CRTR: Updated vertex data for node=0312c02755c6a3475f872be27e71fdb7703a61a25777401a26125e1c772c7e0f6a
2018-03-29 04:49:30.203 [INF] NTFN: Cancelling epoch notification, epoch_id=15092
2018-03-29 04:49:30.203 [INF] HSWC: ChannelLink(393a14959e5e974f2b1af08f50da837baf08a9cd47a78e1c8f91e50dbab8bf13:0) has exited
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x8e8ef7]

goroutine 118602734 [running]:
github.com/lightningnetwork/lnd/lnwallet.(*commitmentChain).tail(...)
        /data/coin/ln/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1051
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).ReceiveNewCommitment(0xc5bb1ed200, 0x95876f9c07166d0e, 0xf027ee841935e65a, 0xca69236234982f72, 0xea563ca1e51eebb5, 0x56846c853f5fe541, 0xf2b3242d3bd13a74, 0x514c87875ae5ef7b, 0xd600a49189a4583f, 0x0, ...)
        /data/coin/ln/go/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:3648 +0xb7
github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).handleUpstreamMsg(0xc5e688a5a0, 0x15b93c0, 0xc4ba8da300)
        /data/coin/ln/go/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:1235 +0xf08
github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).htlcManager(0xc5e688a5a0)
        /data/coin/ln/go/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:921 +0x90e
created by github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).Start
        /data/coin/ln/go/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:391 +0x150

@halseth Good point. Just updated to latest commit. Unfortunately problem still persists and pc has changed

```
commit 6382215346968b69b8b0f5553b7c0bf39e01c550
Merge: 5c42a88 4a200d2
Author: Olaoluwa Osuntokun <laolu32@gmail.com>
Date:   Wed May 23 15:39:30 2018 -0700

```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 **pc=0x8c9a1d**]

goroutine 26430 [running]:
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).logUpdateToPayDesc(0xc4200c2c60, 0xc4213fb510, 0xc4209ace40, 0x4b, 0x2bf2, 0xc4215ae0c0, 0x23d, 0xc420276040, 0xc420276060, 0xc420276080)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1543 +0x55d
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreStateLogs(0xc4200c2c60, 0xc420c4de00, 0xc420c4dec0, 0xc420482540, 0xc4200cee00, 0xc4215ae0c0, 0x0, 0x0)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1732 +0x3a8
github.com/lightningnetwork/lnd/lnwallet.(*LightningChannel).restoreCommitState(0xc4200c2c60, 0xc4207c1808, 0xc4213fb788, 0x0, 0xc4201b9050)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1676 +0x497
github.com/lightningnetwork/lnd/lnwallet.NewLightningChannel(0x10b3280, 0xc42007eea0, 0x7f2443d8bab0, 0xc42007ae40, 0xc4202b2c00, 0xebc676, 0xc, 0xc420040e00)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/lnwallet/channel.go:1371 +0x63e
main.(*peer).loadActiveChannels(0xc4213a0000, 0xc42096c2f8, 0x1, 0x1, 0x2, 0x2)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/peer.go:313 +0x128
main.(*peer).Start(0xc4213a0000, 0xc4208fea00, 0x0)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/peer.go:295 +0x74e
main.(*server).peerConnected(0xc4200aa3c0, 0x10bf620, 0xc4208fea00, 0x0, 0x15b3d00)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/server.go:1487 +0x28a
main.(*server).InboundPeerConnected(0xc4200aa3c0, 0x10bf620, 0xc4208fea00)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/server.go:1563 +0x624
main.(*server).InboundPeerConnected-fm(0x10bf620, 0xc4208fea00)
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/server.go:534 +0x3e
created by github.com/lightningnetwork/lnd/vendor/github.com/roasbeef/btcd/connmgr.(*ConnManager).listenHandler
	/home/d3vt4r/gocode/src/github.com/lightningnetwork/lnd/vendor/github.com/roasbeef/btcd/connmgr/connmanager.go:494 +0x1e6
```

--- a/lnwallet/channel.go
+++ b/lnwallet/channel.go
@@ -4071,6 +4071,7 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 	remoteChainTail := lc.remoteCommitChain.tail().height + 1
 	localChainTail := lc.localCommitChain.tail().height
 
+	source := lc.ShortChanID()
 	chanID := lnwire.NewChanIDFromOutPoint(&lc.channelState.FundingOutpoint)
 
 	// Determine the set of htlcs that can be forwarded as a result of
@@ -4093,11 +4094,14 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 			continue
 		}
 
-		uncommitted := (pd.addCommitHeightRemote == 0 ||
-			pd.addCommitHeightLocal == 0)
-		if pd.EntryType == Add && uncommitted {
-			continue
-		}
+		// For each type of HTLC, we will only consider forwarding it if
+		// both of the remote and local heights are non-zero. If either
+		// of these values is zero, it has yet to be committed in both
+		// the local and remote chains.
+		committedAdd := pd.addCommitHeightRemote > 0 &&
+			pd.addCommitHeightLocal > 0
+		committedRmv := pd.removeCommitHeightRemote > 0 &&
+			pd.removeCommitHeightLocal > 0
 
 		// Using the height of the remote and local commitments,
 		// preemptively compute whether or not to forward this HTLC for
@@ -4105,7 +4109,7 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 		// Settle, Fail, or MalformedFail.
 		shouldFwdAdd := remoteChainTail == pd.addCommitHeightRemote &&
 			localChainTail >= pd.addCommitHeightLocal
-		shouldFwdRmv := remoteChainTail >= pd.removeCommitHeightRemote &&
+		shouldFwdRmv := remoteChainTail == pd.removeCommitHeightRemote &&
 			localChainTail >= pd.removeCommitHeightLocal
 
 		// We'll only forward any new HTLC additions iff, it's "freshly
@@ -4114,8 +4118,7 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 		// don't re-forward any already processed HTLC's after a
 		// restart.
 		switch {
-		case pd.EntryType == Add && shouldFwdAdd:
-
+		case pd.EntryType == Add && committedAdd && shouldFwdAdd:
 			// Construct a reference specifying the location that
 			// this forwarded Add will be written in the forwarding
 			// package constructed at this remote height.
@@ -4128,13 +4131,12 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 			pd.isForwarded = true
 			addsToForward = append(addsToForward, pd)
 
-		case pd.EntryType != Add && shouldFwdRmv:
-
+		case pd.EntryType != Add && committedRmv && shouldFwdRmv:
 			// Construct a reference specifying the location that
 			// this forwarded Settle/Fail will be written in the
 			// forwarding package constructed at this remote height.
 			pd.DestRef = &channeldb.SettleFailRef{
-				Source: lc.ShortChanID(),
+				Source: source,
 				Height: remoteChainTail,
 				Index:  settleFailIndex,
 			}
@@ -4198,8 +4200,6 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 			settleFailUpdates = append(settleFailUpdates, logUpdate)
 		}
 	}
-
-	source := lc.ShortChanID()
 
 	// Now that we have gathered the set of HTLCs to forward, separated by
 	// type, construct a forwarding package using the height that the remote

--- a/lnwallet/channel_test.go
+++ b/lnwallet/channel_test.go
@@ -3943,8 +3943,8 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 	}
 	defer cleanUp()
 
-	// We'll now add two HTLC's from Bob to Alice, then Bob will initiate a
-	// state transition.
+	// We'll now add two HTLC's from Alice to Bob, then Alice will initiate
+	// a state transition.
 	var htlcAmt lnwire.MilliSatoshi = 100000
 	htlc, _ := createHTLC(0, htlcAmt)
 	if _, err := aliceChannel.AddHTLC(htlc, nil); err != nil {
@@ -3975,21 +3975,26 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+
+	// Alice should detect that she doesn't need to forward any HTLC's.
+	fwdPkg, _, _, err := aliceChannel.ReceiveRevocation(bobRevocation)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(fwdPkg.Adds) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.Adds))
+	}
+	if len(fwdPkg.SettleFails) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.SettleFails))
+	}
+
+	// Now, have Bob initiate a transition to lock in the Adds sent by
+	// Alice.
 	bobSig, bobHtlcSigs, err := bobChannel.SignNextCommitment()
 	if err != nil {
 		t.Fatal(err)
-	}
-
-	// Alice should detect that she doesn't need to forward any HTLC's.
-	_, aliceHtlcsToForward, _, err := aliceChannel.ReceiveRevocation(
-		bobRevocation,
-	)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if len(aliceHtlcsToForward) != 0 {
-		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
-			"forward %v htlcs", len(aliceHtlcsToForward))
 	}
 
 	err = aliceChannel.ReceiveNewCommitment(bobSig, bobHtlcSigs)
@@ -4001,15 +4006,19 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Bob on the other hand, should detect that he now has 2 incoming
-	// HTLC's that he can forward along.
-	_, bobHtlcsToForward, _, err := bobChannel.ReceiveRevocation(aliceRevocation)
+	// Bob should now detect that he now has 2 incoming HTLC's that he can
+	// forward along.
+	fwdPkg, _, _, err = bobChannel.ReceiveRevocation(aliceRevocation)
 	if err != nil {
 		t.Fatal(err)
 	}
-	if len(bobHtlcsToForward) != 2 {
+	if len(fwdPkg.Adds) != 2 {
 		t.Fatalf("bob should forward 2 hltcs, instead has %v",
-			len(bobHtlcsToForward))
+			len(fwdPkg.Adds))
+	}
+	if len(fwdPkg.SettleFails) != 0 {
+		t.Fatalf("bob should forward 0 hltcs, instead has %v",
+			len(fwdPkg.SettleFails))
 	}
 
 	// We'll now restart both Alice and Bob. This emulates a reconnection
@@ -4018,18 +4027,20 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unable to restart alice: %v", err)
 	}
+	defer aliceChannel.Stop()
 	bobChannel, err = restartChannel(bobChannel)
 	if err != nil {
 		t.Fatalf("unable to restart bob: %v", err)
 	}
+	defer bobChannel.Stop()
 
 	// With both nodes restarted, Bob will now attempt to cancel one of
 	// Alice's HTLC's.
-	err = bobChannel.FailHTLC(htlc2.ID, []byte("failreason"), nil, nil, nil)
+	err = bobChannel.FailHTLC(htlc.ID, []byte("failreason"), nil, nil, nil)
 	if err != nil {
 		t.Fatalf("unable to cancel HTLC: %v", err)
 	}
-	err = aliceChannel.ReceiveFailHTLC(htlc2.ID, []byte("bad"))
+	err = aliceChannel.ReceiveFailHTLC(htlc.ID, []byte("bad"))
 	if err != nil {
 		t.Fatalf("unable to recv htlc cancel: %v", err)
 	}
@@ -4048,15 +4059,11 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	aliceSig, aliceHtlcSigs, err = aliceChannel.SignNextCommitment()
-	if err != nil {
-		t.Fatal(err)
-	}
 
 	// At this point, Bob receives the revocation from Alice, which is now
 	// his signal to examine all the HTLC's that have been locked in to
 	// process.
-	_, bobHtlcsToForward, _, err = bobChannel.ReceiveRevocation(aliceRevocation)
+	fwdPkg, _, _, err = bobChannel.ReceiveRevocation(aliceRevocation)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -4064,9 +4071,180 @@ func TestLockedInHtlcForwardingSkipAfterRestart(t *testing.T) {
 	// Bob should detect that he doesn't need to forward *any* HTLC's, as
 	// he was the one that initiated extending the commitment chain of
 	// Alice.
-	if len(bobHtlcsToForward) != 0 {
-		t.Fatalf("bob shouldn't forward any htlcs, but has: %v",
-			spew.Sdump(bobHtlcsToForward))
+	if len(fwdPkg.Adds) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.Adds))
+	}
+	if len(fwdPkg.SettleFails) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.SettleFails))
+	}
+
+	// Now, begin another state transition led by Alice, and fail the second
+	// HTLC part-way through the dance.
+	aliceSig, aliceHtlcSigs, err = aliceChannel.SignNextCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = bobChannel.ReceiveNewCommitment(aliceSig, aliceHtlcSigs)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Failing the HTLC here will cause the update to be included in Alice's
+	// remote log, but it should not be committed by this transition.
+	err = bobChannel.FailHTLC(htlc2.ID, []byte("failreason"), nil, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to cancel HTLC: %v", err)
+	}
+	err = aliceChannel.ReceiveFailHTLC(htlc2.ID, []byte("bad"))
+	if err != nil {
+		t.Fatalf("unable to recv htlc cancel: %v", err)
+	}
+
+	bobRevocation, _, err = bobChannel.RevokeCurrentCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Alice should detect that she doesn't need to forward any Adds's, but
+	// that the Fail has been locked in an can be forwarded.
+	_, adds, settleFails, err := aliceChannel.ReceiveRevocation(bobRevocation)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(adds) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(adds))
+	}
+	if len(settleFails) != 1 {
+		t.Fatalf("alice should only forward %d HTLC's, instead wants to "+
+			"forward %v htlcs", 1, len(settleFails))
+	}
+	if settleFails[0].ParentIndex != htlc.ID {
+		t.Fatalf("alice should forward fail for htlcid=%d, instead "+
+			"forwarding id=%d", htlc.ID,
+			settleFails[0].ParentIndex)
+	}
+
+	// We'll now restart both Alice and Bob. This emulates a reconnection
+	// between the two peers.
+	aliceChannel, err = restartChannel(aliceChannel)
+	if err != nil {
+		t.Fatalf("unable to restart alice: %v", err)
+	}
+	defer aliceChannel.Stop()
+	bobChannel, err = restartChannel(bobChannel)
+	if err != nil {
+		t.Fatalf("unable to restart bob: %v", err)
+	}
+	defer bobChannel.Stop()
+
+	// Readd the Fail to both Alice and Bob's channels, as the non-committed
+	// update will not have survived the restart.
+	err = bobChannel.FailHTLC(htlc2.ID, []byte("failreason"), nil, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to cancel HTLC: %v", err)
+	}
+	err = aliceChannel.ReceiveFailHTLC(htlc2.ID, []byte("bad"))
+	if err != nil {
+		t.Fatalf("unable to recv htlc cancel: %v", err)
+	}
+
+	// Have Alice initiate a state transition, which does not include the
+	// HTLCs just readded to the channel state.
+	aliceSig, aliceHtlcSigs, err = aliceChannel.SignNextCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = bobChannel.ReceiveNewCommitment(aliceSig, aliceHtlcSigs)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bobRevocation, _, err = bobChannel.RevokeCurrentCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Alice should detect that she doesn't need to forward any HTLC's, as
+	// the updates haven't been committed by Bob yet.
+	fwdPkg, _, _, err = aliceChannel.ReceiveRevocation(bobRevocation)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(fwdPkg.Adds) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.Adds))
+	}
+	if len(fwdPkg.SettleFails) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(fwdPkg.SettleFails))
+	}
+
+	// Now initiate a final update from Bob to lock in the final Fail.
+	bobSig, bobHtlcSigs, err = bobChannel.SignNextCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = aliceChannel.ReceiveNewCommitment(bobSig, bobHtlcSigs)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	aliceRevocation, _, err = aliceChannel.RevokeCurrentCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Bob should detect that he has nothing to forward, as he hasn't
+	// received any HTLCs.
+	fwdPkg, _, _, err = bobChannel.ReceiveRevocation(aliceRevocation)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(fwdPkg.Adds) != 0 {
+		t.Fatalf("bob should forward 4 hltcs, instead has %v",
+			len(fwdPkg.Adds))
+	}
+	if len(fwdPkg.SettleFails) != 0 {
+		t.Fatalf("bob should forward 0 hltcs, instead has %v",
+			len(fwdPkg.SettleFails))
+	}
+
+	// Finally, have Bob initiate a state transition that locks in the Fail
+	// added after the restart.
+	aliceSig, aliceHtlcSigs, err = aliceChannel.SignNextCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = bobChannel.ReceiveNewCommitment(aliceSig, aliceHtlcSigs)
+	if err != nil {
+		t.Fatal(err)
+	}
+	bobRevocation, _, err = bobChannel.RevokeCurrentCommitment()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// When Alice receives the revocation, she should detect that she
+	// can now forward the freshly locked-in Fail.
+	_, adds, settleFails, err = aliceChannel.ReceiveRevocation(bobRevocation)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(adds) != 0 {
+		t.Fatalf("alice shouldn't forward any HTLC's, instead wants to "+
+			"forward %v htlcs", len(adds))
+	}
+	if len(settleFails) != 1 {
+		t.Fatalf("alice should only forward one HTLC, instead wants to "+
+			"forward %v htlcs", len(settleFails))
+	}
+	if settleFails[0].ParentIndex != htlc2.ID {
+		t.Fatalf("alice should forward fail for htlcid=%d, instead "+
+			"forwarding id=%d", htlc2.ID,
+			settleFails[0].ParentIndex)
 	}
 }
 

